### DISCUSSION TITLE
Running the webui on Mac OS

### DISCUSSION THREAD

**Discussion by OrT7** [lllyasviel/stable-diffusion-webui-forge#2105] [Q&A] [answered]
How can I run the webui on my M1 Mac?
I did manage to use it on my Windows PC, but I am not sure how to run the update.bat and the webui.bat on my Mac
I don't want to use a VM for that and I am sure there is a way to do it.
Do you know how to do it? Please help

**Comment by jswag245** [comment, marked as answer]
I'd imagine it's somewhat similar to this method for Automatic:
https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Installation-on-Apple-Silicon

**Comment by jswag245** [comment]
Can you confirm that this solution currently works for you on the latest version of Forge?

**Reply by reekes** [reply]
I've been running Automatic 1111 for years, on a Mac. Currently on the M2 Ultra. It's been a struggle since there's so little support for Mac users.

I installed Forge without any modifications. So far I can only generate blank/black images. I want to use Flux, since Automatic 1111 doesn't support it. After too much searching, I haven't found anyone successfully running Forge on a Mac that can document how to get it working. Also, I don't see any performance improvements with Forge on a Mac. For me, I just want to take a test drive and get access to Flux.

As an example of my frustration. I tried something very basic, get the dark theme. I tried uncommenting `export COMMANDLINE_ARGS` in webui-user.sh to pass `--theme=dark` and that causes Forge to fail at launch. It's the exact same thing I have in Automatic 1111 which works. I can pass `--theme=dark` on the cmd line myself, but haven't found what magic would work within the scripts. This is something so basic yet it fails, and I don't find any docs on such things.

Same with the important option to share all gigs of models in Automatic 1111 (`--forge-ref-a1111-home`). Nothing I've tried using the "right" approach of adding this to webui-user.sh has worked. Typing these options on the cmd line works, but not if you add it to the startup scripts as expected. Makes me think nobody is testing this code on a Mac.

FYI - in Automatic 1111 here's what I'm running below. I get "decent" performance after **a lot** of tweaking. My stupid expensive Mac Studio Pro performing at the speed of a cheap Windows laptop, costing about 1/10th the price. After a few days of trying to use Stable Diffusion on a Mac, I just get frustrated and exhausted. Then I try again in six month. Only to spend many more days trying to get it working, again, and then growing frustrated and exhausted.

Automatic 1111 version: [v1.10.0-RC-60-g8b3d98c5]
python: 3.10.15
torch: 2.5.1
xformers: N/A
gradio: 3.41.2

Performance results using SD WebUI Benchmark:

3.64 it/s
4.15 it/s
4.56 it/s

**Comment by MigCurto** [comment]
I run Flux on Forge with Mac M2 Ultra without any problems , I do have the 192GB version and run pytorch nightly version (updated last week)so not sure you can translate it, but just edited my webui-user.sh like this:

```
export COMMANDLINE_ARGS="--skip-torch-cuda-test --theme dark --all-in-fp16 --attention-pytorch  --use-cpu interrogate --no-download-sd-model "

export TORCH_COMMAND="pip install --pre torch torchvision --index-url https://download.pytorch.org/whl/nightly/cpu"
venv_dir="venv-torch-nightly"
```
<img width="1079" alt="image" src="https://github.com/user-attachments/assets/6bde0b18-0690-4668-ada7-50dc352ab0e7">


Forge scripts take care of installation and setup of pytorch and venvs, just need to edit the webui-user.sh
I do keep the stable pytorch venv in hand in case one of the nightlys goes bad, its just a matter of comment/uncomment with #'s

It runs Flux about the same speed as Comfy, altought for some reason I cant get pass some resolutions while upscaling with ultimateSD upscale in Comfy but got no problems with Forge.

The thing is for Flux AFAIK Macs have to use fp16 text encoder (apparently they dont support fp8_e4m3fn) , wich slows down things a bit:
<img width="1156" alt="image" src="https://github.com/user-attachments/assets/88541e07-45e2-4ffd-97bd-30eb30a60c75">

Like someone said: Flux is a fat cow, but a juicy one...

If you want speed use SDXL Lightning or Turbo.

**Reply by reekes** [reply]
Thanks for the reply @MigCurto, I really do appreciate it.

I did a new/clean pull of Forge. Added the two changes for `COMMANDLINE_ARGS` and `TORCH_COMMAND` as you suggested. Stuff I had done previously just a couple days ago, in yet another attempt to get a working system.

Since I do have Automatic 1111 already full installed/configured, I added `--forge-ref-a1111-home`

Launched and waited for Forge to finish the install.

Got this in the console:

```
Legacy Preprocessor init warning: Unable to install insightface automatically. Please try run pip install insightface manually.
```

Do I need that? Forge didn't install it, but seems to be complaining. If I need it, what's the best way to install it. I'm using homebrew. Should I add it to my venv? Instead of `pip` I have `pip3`. Maintaining python and all the libs needed by Stable Diffusion on macOS is poorly documented. Python alone is a PITA, and Automatic 1111 doesn't seem compatible with the current python.

I've never found a way to avoid this warning, since there is no CUDA on our Macs. I would expect this option to avoid it `--skip-torch-cuda-test`

```
Warning: caught exception 'Torch not compiled with CUDA enabled', memory monitor disabled
```

Finally I get to the main screen, and select a SDXL checkpoint entering only "cat" as a prompt. No other changes. Image size is 512x640.

Then I get this in the console:

```
Moving model(s) has taken 46.77 seconds
```

Why does it take a minute to load a model? I've seen this same result with several attempts.

**Surprise!** This time I didn't get a blank/black image. Hmm

So I enter a more realistic prompt, and generate. Again, it takes another minute to start. It's using the same model. Why is this so slow? Running an XYZ plot with a dozen models is promising to be really painful.

```
Moving model(s) has taken 16.56 seconds
[Unload] Trying to free 7390.16 MB for mps with 0 models keep loaded ... Current free memory is 63407.91 MB ... Done.
All loaded to GPU.
Moving model(s) has taken 35.02 seconds

```
At least this time I got an image. Not a good one, but something. My previous setup must have a setting or extension causing a problem.

I only use a few extensions, ones I found vital to my workflow. There are a few settings I've changed in Automatic 1111 (most just to make things work on my Mac, but others I did for performance). I duplicated those previously which may have caused a problem in Forge. But it's also odd since there were no warnings/errors in the console, just a black image. One or more of those might be a problem. I'll have to start the long process of testing one-by-one to see if I can get Forge working where I need it.

If and when I get there, then I'll attempt to do what I'm actually trying which is to get Flux working.

Thanks again for prompting me to give it another go. I'll reply after I find anything or run into another blocker.

**Comment by MigCurto** [comment]
`Legacy Preprocessor init warning: Unable to install insightface automatically. Please try run pip install insightface manually.`

That's not "normal" but I remember I got that too, there's a bug somewhere on installs or dependencies that prevent insightface to be installed.But that's only necessary for some extra features and its easy to solve: locate and deploy the file to activate the venv and install that python module in terminal:  " source /path to venv/bin/activate

In my case (personalized venv name but yours should be similar) would be something like:

```
source /Users/miguel/stable-diffusion-webui-forge/venv-torch-nightly/bin/activate 
pip install insightface
```
`Warning: caught exception 'Torch not compiled with CUDA enabled', memory monitor disabled`
Perfectly normal in Macs, no CUDA.

Your image problems could have to do with sampler/schedule chosen, Flux is very picky, also CFG must be kept at 1 (for Both Flux or Turbo/Lightning SDXL)

Regarding performace: SDXL Turbo or Lightning are fast on a M2 Ultra, around 1.50/1.75its/s, can get 4 images under 30sec:

<img width="1771" alt="image" src="https://github.com/user-attachments/assets/a4d68020-0e08-4977-9a7d-aa0ebc8cc5c5">

Flux not so much, I usually get 1 image in around 1.15min/1.30min:
<img width="1779" alt="image" src="https://github.com/user-attachments/assets/7ee1c6c6-c32f-4963-92d5-2d6656604e9e">


Also, found no real benefits in using GUFF as time wise is slower and I have memory to spare (like many Mac ARM users with unified memory) . And bnb-nf4 flavours AFAIK cant be deployed in Macs because Bits and Bytes hasnt been ported to Mac.So I just use regular Dev , you could try Dev-f8 but I feel it kinds of defeats the point for me running Flux (better quality) and Schnell is a completely different thing.

**Reply by reekes** [reply]
@MigCurto I've got Forge running on my Mac M2 Ultra. Been running through various SDXL models and testing. So far it works. I haven't tried Flux yet. I was hoping for good news with SDXL.

Unfortunately, I find that Forge is slower doing the same renderings than Automatic1111. The [benchmark test](https://github.com/vladmandic/sd-extension-system-info) using literally the same model gives these results.

Automatic1111: 3.64 / 3.95 / 4.44
Forge: 2.69 / 3.01 / 3.28

I don't know what options might make Forge run faster, but it's performance out of the box is not good. I tried some of the optimizations I have working well from Automatic1111, but this caused Forge to fail or hang. I would have expected the default install to be better than what I have already working in Automatic1111.

I don't care for the Turbo or Lightning models as they compromise quality.

**Reply by MigCurto** [reply]
I only use Flux these days, not sure about the status today but A1111 wasnt an option for Flux a while ago.
I got pretty much the same speeds in Comfy/Forge, the arguments I use (not extensively tested) are:

--all-in-fp16 --attention-pytorch --use-cpu interrogate

Not sure if it helps on SDXL thought, but give it a try.
EDIT: I use pytorch nightly.
EDIT2: Don't quote my on this but I think the real problem with Flux on Macs is that AFAIK  the are no ARM chips capable of using fp8_e4m3fn so we take a bigger hit on speed.
